### DISCUSSION TITLE
Instantiation of values of structured data types

### DISCUSSION THREAD

**Discussion by DavidKrGH** [FreeOpcUa/opcua-asyncio#1271] [Q&A]
Hello to all OPC UA interested. I have a question regarding Structured Data Type variables. The larger goal is to browse an existing OPC UA server through a client and export certain data structures as XML. Afterwards the XML shall be imported by a server. An existing server is to be cloned in this way. So much for the context.

The original server uses its own enums as well as structures that are used to create object types. The described elements are shown in the pictures. The variable "OperationModeInfo" has a type definition that points to the object type "OpconOperationModeInfo". As you can see in the pictures, there is a structure for this, in which the fields are defined. 

<img width="173" alt="OperationModeInfo" src="https://user-images.githubusercontent.com/86784433/229441973-e275fa01-0867-45fd-bc8e-a1a88dad4363.png">
<img width="184" alt="OpconOperationModeInfo" src="https://user-images.githubusercontent.com/86784433/229442068-2636a05a-c12b-4b81-82aa-2e15579fc495.png">
<img width="427" alt="OperationModeInfo_attr" src="https://user-images.githubusercontent.com/86784433/229442110-847edde4-18c6-48ab-9d92-3b5f2de2c429.png">
<img width="455" alt="2023-03-31_13h26_30" src="https://user-images.githubusercontent.com/86784433/229444050-7f7c9f4e-9eed-4699-9f8d-06d0379dcfbf.png">

On the original server, the individual values of "OperationModeInfo" are each accessible via their own NodeId and are listed as variables. However, when such an object is created on the FreeOpcUa server, this is not the case. The variables are only accessible via "OperationModeInfo" itself (cf. screenshots). 
<img width="1209" alt="Code" src="https://user-images.githubusercontent.com/86784433/229443945-9b4b9ce2-4f75-4d44-9044-6cd16e3cdf32.png">
<img width="166" alt="OperationModeInfo_New" src="https://user-images.githubusercontent.com/86784433/229443970-f1204f80-c8f4-4a1e-9917-7c31657b3cb4.png">
<img width="427" alt="OperationModeInfo_New_attr" src="https://user-images.githubusercontent.com/86784433/229444017-e6ccfba1-dd17-488e-b42f-b2a72ccfb49b.png">

Is there a way to create the identical data structure with individually reachable values in FreeOpcUa?

Many thanks for your help.

**Comment by schroeder-** [comment]
Has the node a reference of HasStructuredComponent (https://reference.opcfoundation.org/Core/Part3/v105/docs/7.22)?

This feature is currently not support.

**Reply by DavidKrGH** [reply]
Thanks for the quick reply.

There is a "HasLocalVars" reference. This is a subtype of the "HasComponentReference". I assume that this corresponds to a "HasStructuredComponent" reference.

Is there perhaps a workaround to keep the values of the main component syncronous with those of the subcomponents?

**Reply by schroeder-** [reply]
You have to create a subscription to the variable and update the child values on update. 

``` Python
class SubHandler:
    """
    Subscription Handler. To receive events from server for a subscription
    """

    async def datachange_notification(self, node, val, data):
        dic_v = asdict(val)
        for name, value in dic_v:
              v = await node.get_child(name)
              await v.set_value(value)


handler = SubHandler()
sub = await server.create_subscription(500, handler)
await sub.subscribe_data_change(node)
```
